base/error.jl

```diff
@@ -49,3 +49,47 @@ macro assert(ex, msgs...)
     end
     :($(esc(ex)) ? $(nothing) : throw(Main.Base.AssertionError($msg)))
 end
+
+
+"""
+    retry(f, [condition], n=3; max_delay=10)
+
+Returns a lambda that retries function `f` up to `n` times in the
+event of an exception. If `condition` is a `Type` then retry only
+for exceptions of that type. If `condition` is a function
+`cond(::Exception) -> Bool` then retry only if it is true.
+
+e.g. `retry(http_get, e->e.status == "503")(url)` or `retry(read, UVError)(io)`.
+"""
+function retry(f::Function, condition::Function=e->true, n::Int=3; max_delay=10)
+    (args...) -> begin
+        delay = 0.05
+        for i = 1:n
+            try
+                return f(args...)
+            catch e
+                if i == n || try condition(e) end != true
+                    rethrow(e)
+                end
+            end
+            sleep(delay * (0.8 + (rand() * 0.4)))
+            delay = min(max_delay, delay * 5)
+        end
+    end
+end
+
+retry(f::Function, condition::Type, n::Int=3) = retry(f, e->isa(e, condition), n)
```

Comment:
`retry(f::Function, condition::Type, n::Int=3; kwargs) = retry(f, e->isa(e, condition), n; kwargs...)` will allow supporting `max_delay` in both signatures.

`n::Int=3` should be `n=NUM_DEFAULT_RETRY` so that the value is defined in only one place.